File: fixtures/code_corpus/javascript/middleware.js
/**
 * Express middleware utilities
 */

class Middleware {
  static errorHandler(err, req, res, next) {
    console.error('Error:', err);

    if (err.type === 'entity.parse.failed') {
      return res.status(400).json({
        error: 'Invalid JSON',
        message: err.message
      });
    }

    if (err.name === 'ValidationError') {
      return res.status(400).json({
        error: 'Validation Error',
        message: err.message
      });
    }

    res.status(500).json({
      error: 'Internal Server Error',
      message: process.env.NODE_ENV === 'development' ? err.message : undefined
    });
  }

  static requestLogger(req, res, next) {
    const start = Date.now();

    res.on('finish', () => {
      const duration = Date.now() - start;
      console.log(`${req.method} ${req.path} ${res.statusCode} ${duration}ms`);
    });

    next();
  }

  static tenantId(req, res, next) {
    const tenantId = req.headers['x-tenant-id'] || req.headers['X-Tenant-ID'];

    if (!tenantId) {
      return res.status(400).json({
        error: 'Missing Tenant ID',
        message: 'X-Tenant-ID header is required'
      });
    }

    req.tenantId = tenantId;
    next();
  }

  static async auth(req, res, next) {
    const token = req.headers.authorization?.replace('Bearer ', '');

    if (!token) {
      return res.status(401).json({
        error: 'Unauthorized',
        message: 'Missing authentication token'
      });
    }

    try {
      // Verify token (implementation depends on auth system)
      req.user = { id: 'user_123' }; // Placeholder
      next();
    } catch (error) {
      res.status(401).json({
        error: 'Unauthorized',
        message: 'Invalid token'
      });
    }
  }

  static cors(req, res, next) {
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
    res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-Tenant-ID');

    if (req.method === 'OPTIONS') {
      return res.sendStatus(204);
    }

    next();
  }
}

module.exports = Middleware;

File: backend/tests/fixtures/block_z/code_files/middleware.py
"""Middleware for request/response handling."""
from typing import Callable
import time

def auth_middleware(get_response: Callable):
    """Authentication middleware."""
    def middleware(request):
        token = request.headers.get("Authorization")
        if not token or not token.startswith("Bearer "):
            raise Exception("Unauthorized: Missing or invalid token")
        
        # Verify token (simplified)
        request.user = {"id": "user123", "email": "user@example.com"}
        response = get_response(request)
        return response
    
    return middleware

def logging_middleware(get_response: Callable):
    """Logging middleware."""
    def middleware(request):
        start_time = time.time()
        print(f"Request: {request.method} {request.path}")
        
        response = get_response(request)
        
        elapsed = time.time() - start_time
        print(f"Response: {response.status_code} ({elapsed:.3f}s)")
        return response
    
    return middleware

def cors_middleware(get_response: Callable):
    """CORS middleware."""
    def middleware(request):
        response = get_response(request)
        response.headers["Access-Control-Allow-Origin"] = "*"
        response.headers["Access-Control-Allow-Methods"] = "GET, POST, PUT, DELETE, OPTIONS"
        response.headers["Access-Control-Allow-Headers"] = "Content-Type, Authorization"
        return response
    
    return middleware

def rate_limit_middleware(get_response: Callable, max_requests: int = 100):
    """Rate limiting middleware."""
    request_counts = {}
    
    def middleware(request):
        client_ip = request.META.get("REMOTE_ADDR")
        current_time = int(time.time() / 60)  # Per minute
        key = f"{client_ip}:{current_time}"
        
        count = request_counts.get(key, 0)
        if count >= max_requests:
            raise Exception("Rate limit exceeded")
        
        request_counts[key] = count + 1
        return get_response(request)
    
    return middleware
